ql/src/java/org/apache/hadoop/hive/ql/optimizer/GroupingSetOptimizer.java

```diff
@@ -0,0 +1,362 @@
+      while (true) {
+        if (curOp instanceof ReduceSinkOperator) {
+          rs = (ReduceSinkOperator) curOp;
+          break;
+        }
+
+        if (curOp.getParentOperators() == null) {
+          break;
+        }
+
+        if (curOp.getParentOperators().size() == 1) {
+          curOp = curOp.getParentOperators().get(0);
+        } else if (curOp instanceof AbstractMapJoinOperator) {
+          MapJoinDesc desc = ((AbstractMapJoinOperator<?>) curOp).getConf();
+          curOp = curOp.getParentOperators().get(desc.getPosBigTable());
+        } else {
+          break;
+        }
+      }
+
+      if (rs == null) {
+        // There is no partitioning followed by this parentOp. Continue optimization.
+        return true;
+      }
+
+      if (rs.getConf().getPartitionCols() != null && rs.getConf().getPartitionCols().size() > 0) {
+        // This rs might be irrelevant to the target GroupBy operator. For example, the following query:
+        //   SELECT a, b, sum(c) FROM (SELECT a, b, c FROM tbl DISTRIBUTE BY c) z GROUP BY rollup(a, b)
+        // won't be optimized although 'DISTRIBUTE BY c' is irrelevant to the key columns of GroupBy.
+        LOG.debug("Skip grouping-set optimization in order not to introduce possibly redundant shuffle.");
+        return false;
+      } else {
+        // No partitioning. Continue optimization.
+        return true;
+      }
+    }
+
+    private String selectPartitionColumn(GroupByOperator gby, Operator<?> parentOp) {
+      if (!(parentOp.getSchema() != null && parentOp.getSchema().getSignature() != null)) {
+        LOG.debug("Skip grouping-set optimization as the parent operator {} does not provide signature",
+            parentOp);
+        return null;
+      }
+
+      if (!(parentOp.getStatistics() != null && parentOp.getStatistics().getColumnStats() != null)) {
+        LOG.debug("Skip grouping-set optimization as the parent operator {} does not provide statistics",
+            parentOp);
+        return null;
+      }
+
+      if (parentOp.getStatistics().getNumRows() > gby.getStatistics().getNumRows()) {
+        LOG.debug("Skip grouping-set optimization as the parent operator {} emits more rows than {}",
+            parentOp, gby);
+        return null;
+      }
+
+      List<String> colNamesInSignature = new ArrayList<>();
+      for (ColumnInfo pColInfo: parentOp.getSchema().getSignature()) {
+        colNamesInSignature.add(pColInfo.getInternalName());
+      }
+
+      List<Integer> groupingSetKeys = listGroupingSetKeyPositions(gby.getConf().getListGroupingSets());
+      Set<String> candidates = new HashSet<>();
+      for (Integer groupingSetKeyPosition: groupingSetKeys) {
+        ExprNodeDesc key = gby.getConf().getKeys().get(groupingSetKeyPosition);
+
+        if (key instanceof ExprNodeColumnDesc) {
+          candidates.add(((ExprNodeColumnDesc) key).getColumn());
+        }
+      }
+      candidates.retainAll(colNamesInSignature);
```

Comment:
I am not understanding why we can put all of the parent operator.

Reply:
In the above code, we gather the parent columns that are used for groupingSet. The for-loop gathers all of the such columns, and I call retainAll() just to ensure that the columns gathered from groupingSet actually exist in parent operator.

Reply:
Ah, ok. Please ignore. I thought I saw not `retainAll` but `addAll`. My bad